rfcs/text/responsiveGrid.md

```diff
@@ -0,0 +1,407 @@
+1. Gutters are calculated automatically based on a the number of `GridItems`.
+1. Can be re-ordered on the page by viewport dimensions.
+1. Can be styled to look like a container.
+1. Can be passed a range of values for padding.
+
+### Gutters
+Gutters for the [`GridItem`](#GridItem) are set automatically in the CSS, following the values in this table.
+
+| Breakpoints                | Device Type  | Gutters |
+|----------------------------|--------------|---------|
+| Extra Small (320 to 599px) | Smart Phones | 16px    |
+| Small (600 to 959px)       | Portrait Tablets | 16px |
+| Medium (960 to 1259px) | Landscape Tablets & Low-Res Laptops | 24px |
+| Large (1260 to 1920px) | High-Res Laptops & Monitors | 24px |
+| Extra Large (1921px and above) | Ultra High-Res Monitors |40px |
+
+## Code example:
+Below is an example of how a Grid could be included as a component and used to implement the [`proposed solution used in the CSS comparison`](#comparison-methodology).
+```js
+import React from "react";
+import { ThemeProvider } from "styled-components";
+import { mintTheme } from "carbon-react/lib/style/themes";
+import "carbon-react/lib/utils/css";
+import { startRouter } from "carbon-react/lib/utils/router";
+import AppWrapper from "carbon-react/lib/components/app-wrapper";
+import { GridContainer, GridItem } from "./components/grid";
+
+const App = () => {
+  const item1_900 = {
+    maxWidth: 900,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 9,
+    rowStart: 2,
+    rowEnd: 2
+  };
+  const item1_1300 = {
+    maxWidth: 1300,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 13,
+    rowStart: 1,
+    rowEnd: 1  
+  };
+  const item1_1500 = {
+    maxWidth: 1500,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 7,
+    rowStart: 1,
+    rowEnd: 1
+  }
+  const responsiveItem1Settings = [
+    item1_1300,
+    item1_900,
+    item1_1500
+  ];
+
+  const item2_900 = {
+    maxWidth: 900,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 9,
+    rowStart: 3,
+    rowEnd: 3
+  };
+  const item2_1300 = {
+    maxWidth: 1300,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 13,
+    rowStart: 2,
+    rowEnd: 2
+  };
+  const item2_1500 = {
+    maxWidth: 1500,
+    unitType: 'px',
+    colStart: 7,
+    colEnd: 13,
+    rowStart: 1,
+    rowEnd: 1
+  };
+  const responsiveItem2Settings = [
+    item2_1300,   
+    item2_900,
+    item2_1500
+  ];
+
+  const item3_900 = {
+    maxWidth: 900,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 9,
+    rowStart: 1,
+    rowEnd: 1
+  };
+  const item3_1300 = {
+    maxWidth: 1300,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 13,
+    rowStart: 3,
+    rowEnd: 3
+  };
+  const item3_1500 = {
+    maxWidth: 1500,
+    unitType: 'px',
+    colStart: 1,
+    colEnd: 13,
+    rowStart: 2,
+    rowEnd: 2
+  };
+  const responsiveItem3Settings = [
+    item3_1300,
+    item3_900,
+    item3_1500
+  ];
+
+  return (
```

Comment:
How about this interface:

```js
<GridContainer breakpoints={[900, 1300, 1500?]}>
  <GridItem colStart={1} colEnd={[9,13,7]} rowStart={[2,1,1]} rowEnd={[2,1,1]}>
    cell 1
  </GridItem>
  <GridItem colStart={[1,1,7]} colEnd={[9,13,13]} rowStart={[3,2,1]} rowEnd={[3,2,1]}> 
    cell 2
  </GridItem>
  <GridItem colStart={1} colEnd={[9,13,13]} rowStart={[1,3,2]} rowEnd={[1,3,2]}>
    cell 3
  </GridItem>
</GridContainer>
```

Reply:
I experimented with this interface and ran into issues with defaults and overides. Following discussion with @jamime we'd like to preserve the proposed interface as a configurable object. This allows us to omit breakpoints and/or css properties for individual GridItems. I've added further code fragments to illustrate this.